src/main/java/com/google/cloud/tools/appengine/cloudsdk/serialization/CloudSdkVersion.java

```diff
@@ -20,71 +20,119 @@
 import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableList;
 
-import java.util.ArrayList;
 import java.util.List;
 import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /**
- * Represents the version of the Cloud SDK. Loosely follows the semantic versioning spec
- * (semver.org) with a few exceptions to make it more flexible:
- *
- * <ul>
- *   <li>Versions can have one or more numeric version components, instead of MAJOR.MINOR.PATCH.
- *   These numeric version components are compared for ordering and equality testing.</li>
- *   <li>Any pre-release or build numbers are ignored for ordering and equality testing.</li>
- * </ul>
+ * Represents a Version of the Cloud SDK, which follows the Semantic Version 2.0.0 spec. See
+ * <a href="http://semver.org/spec/v2.0.0.html">http://semver.org/spec/v2.0.0.html</a> for more
+ * detail.
  */
 public class CloudSdkVersion implements Comparable<CloudSdkVersion> {
 
-  private static final char BUILD_SEPARATOR = '+';
-  private static final char PRERELEASE_SEPARATOR = '-';
+  private static final Pattern SEMVER_PATTERN = Pattern.compile(getSemVerRegex());
 
-  private final List<Integer> versionComponents;
   private final String version;
 
+  private final int majorVersion;
+  private final int minorVerion;
+  private final int patchVersion;
+  private final CloudSdkVersionPreRelease preRelease;
+  private final String build;
+
   /**
-   * Constructs a CloudSdkVersion from a version string.
-   * @param version a non-null, nonempty string of the form "\d+(\.\d+)*[+-].*".
-   * @throws IllegalArgumentException if the string cannot be parsed
+   * Constructs a new CloudSdkVersion.
+   *
+   * @param version the semantic version string
+   * @throws IllegalArgumentException if the argument is not a valid semantic version string
    */
   public CloudSdkVersion(String version) throws IllegalArgumentException {
     Preconditions.checkArgument(!Strings.isNullOrEmpty(version));
 
+    Matcher matcher = SEMVER_PATTERN.matcher(version);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException(
+          String.format("Pattern \"%s\" is not a valid CloudSdkVersion.", version));
+    }
+
+    majorVersion = Integer.parseInt(matcher.group("major"));
+    minorVerion = Integer.parseInt(matcher.group("minor"));
+    patchVersion = Integer.parseInt(matcher.group("patch"));
+
+    preRelease = matcher.group("prerelease") != null
+        ? new CloudSdkVersionPreRelease(matcher.group("prerelease")) : null;
+    build = matcher.group("build");
+
     this.version = version;
-    this.versionComponents = parseVersionComponents(version);
+  }
+
+  private static String getSemVerRegex() {
```

Comment:
Yep, this is similar to the regex that the Cloud SDK uses for version comparison.